types/defines/vectorize.d.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2022-2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+/**
+ * Additional information to associate with a vector.
+ */
+type VectorizeVectorMetadata =
+  | string
+  | number
+  | boolean
+  | string[]
+  | Record<string, string | number | boolean | string[]>;
+
+type VectorFloatArray = Float32Array | Float64Array;
+
+interface VectorizeError {
+  code?: number;
+  error: string;
+}
+
+/**
+ * A pre-configured list of known models.
+ * These can be supplied in place of configuring explicit dimensions.
+ */
+type VectorizeKnownModel =
+  | "openapi-text-embedding-ada-002"
+  | "workers-ai/bge-small-en"
+  | "cohere/embed-multilingual-v2.0"
+
+/**
+ * Supported distance metrics for an index.
+ * Distance metrics determine how other "similar" vectors are determined.
+ */
+type VectorizeDistanceMetric =
+  | "euclidean"
+  | "cosine"
+  | "dot-product"
+
+interface VectorizeQueryOptions {
+  topK?: number;
+  namespace?: string;
+  returnVectors?: boolean;
+}
+
+/**
+ * Information about the configuration of an index.
+ */
+interface VectorizeIndexConfig {
+  dimensions?: number;
+  knownModel?: VectorizeKnownModel;
+  metric: VectorizeDistanceMetric;
```

Comment:
Should `metric` also be optional?

You must specify `knownModel` (or `preset`) _or (dimensions + metric)

Reply:
Unions to the rescue. Ship it.